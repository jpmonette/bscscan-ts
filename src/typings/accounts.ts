import { BEP20, ERC721 } from ".";

type Sort = "asc" | "desc";
type Tag = "latest" | string;
type BlockType = "blocks" | string;

export type Pagination = {
  sort?: Sort;
  startblock?: number;
  endblock?: number;
  page?: number;
  offset?: number;
};

export type GetBalanceRequest = {
  address: string;
  tag?: Tag;
};

export type GetBalanceResponse = string;

export type GetBalanceMultiRequest = {
  address: string[];
  tag?: Tag;
};

export type GetBalanceMultiResponse = Array<GetBalanceMultiResponseItem>;

export type GetBalanceMultiResponseItem = {
  account: string;
  balance: string;
};

export type GetTxListRequest = Pagination & {
  address: string;
};

export type GetTxListResponse = Array<GetTxListResponseItem>;

export type GetTxListResponseItem = {
  blockHash: string;
  blockNumber: string;
  confirmations: string;
  contractAddress: string;
  cumulativeGasUsed: string;
  from: string;
  gas: string;
  gasPrice: string;
  gasUsed: string;
  hash: string;
  input: string;
  isError: string;
  nonce: string;
  timeStamp: string;
  to: string;
  transactionIndex: string;
  txreceipt_status: string;
  value: string;
};

export type GetTxListInternalRequest = GetTxListRequest;

export type GetTxListInternalResponse = Array<GetTxListInternalResponseItem>;

export type GetTxListInternalResponseItem = {
  blockNumber: string;
  contractAddress: string;
  errCode: string;
  from: string;
  gas: string;
  gasUsed: string;
  hash: string;
  input: string;
  isError: string;
  timeStamp: string;
  to: string;
  traceId: string;
  type: string;
  value: string;
};

export type GetTokenTxRequest = GetTxListRequest & {
  contractaddress?: string;
};

export type GetTokenTxResponse = Array<BEP20>;

export type GetTokenNFTTxRequest = Pagination & {
  address: string;
};

export type GetTokenNFTTxResponse = Array<ERC721>;

export type GetMinedBlocksRequest = Pagination & {
  address: string;
  blocktype?: BlockType;
};

export type GetMinedBlocksResponse = Array<GetMinedBlocksResponseItem>;

export type GetMinedBlocksResponseItem = {
  blockNumber: string;
  timeStamp: string;
  blockReward: string;
};
