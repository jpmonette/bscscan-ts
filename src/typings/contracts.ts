export type GetAbiRequest = {
  // the contract address that has a verified source code
  address: string;
};

export type GetAbiResponse = string;

export type GetSourceCodeRequest = {
  // the contract address that has a verified source code
  address: string;
};

export type GetSourceCodeResponse = Array<GetSourceCodeResponseItem>;

export type GetSourceCodeResponseItem = {
  SourceCode: string;
  ABI: string;
  ContractName: string;
  CompilerVersion: string;
  OptimizationUsed: string;
  Runs: string;
  ConstructorArguments: string;
  EVMVersion: string;
  Library: string;
  LicenseType: string;
  Proxy: string;
  Implementation: string;
  SwarmSource: string;
};

export type VerifySourceCodeRequest = {
  contractAddress: string;
  sourceCode: string;
  // solidity-single-file (default) or solidity-standard-json-input (for std-input-json-format support)
  codeFormat?: "solidity-single-file" | "solidity-standard-json-input";
  contractName: string;
  //Â see https://bscscan.com/solcversions for list of support versions
  compilerVersion: string;
  // 0 = No Optimization, 1 = Optimization used (applicable when codeformat=solidity-single-file)
  optimizationUsed: "0" | "1";
  // set to 200 as default unless otherwise  (applicable when codeformat=solidity-single-file)
  runs: number;
  constructorArguments?: string;
  // leave blank for compiler default, homestead, tangerineWhistle, spuriousDragon, byzantium, constantinople, petersburg, istanbul (applicable when codeformat=solidity-single-file)
  evmVersion?: string;
  // id codes 1-12 where 1=No License .. 12=Apache 2.0, see https://bscscan.com/contract-license-types
  licenseType: string;
  libraryName1?: string;
  libraryAddress1?: string;
  libraryName2?: string;
  libraryAddress2?: string;
  libraryName3?: string;
  libraryAddress3?: string;
  libraryName4?: string;
  libraryAddress4?: string;
  libraryName5?: string;
  libraryAddress5?: string;
  libraryName6?: string;
  libraryAddress6?: string;
  libraryName7?: string;
  libraryAddress7?: string;
  libraryName8?: string;
  libraryAddress8?: string;
  libraryName9?: string;
  libraryAddress9?: string;
  libraryName10?: string;
  libraryAddress10?: string;
};

export type VerifySourceCodeResponse = string;

export type VerifyProxyContractRequest = {
  address: string;
  expectedImplementation?: string;
};

export type VerifyProxyContractResponse = string;

export type CheckProxyVerificationRequest = {
  guid: string;
};

export type CheckProxyVerificationResponse = string;
