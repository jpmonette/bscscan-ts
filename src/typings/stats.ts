export type GetBNBTotalSupplyResponse = string;

export type GetValidatorsResponse = Array<GetValidatorsResponseItem>;

export type GetValidatorsResponseItem = {
  validatorAddress: string;
  validatorName: string;
  validatorStatus: string;
  validatorVotingPower: string;
  validatorVotingPowerProportion: string;
};

export type GetBNBLastPriceResponse = {
  ethbtc: string;
  ethbtc_timestamp: string;
  ethusd: string;
  ethusd_timestamp: string;
};
