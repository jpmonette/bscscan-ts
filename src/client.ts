import { Accounts } from "./modules/accounts";
import { Contracts } from "./modules/contracts";
import { Stats } from "./modules/stats";
import { Transactions } from "./modules/transactions";
import { APIResponse, BscScanOptions } from "./typings";

/**
 * The default production API base url
 */
const BASE_URL = "https://api.bscscan.com/api?";

/**
 * A BscScan API client
 *
 * @class BscScan
 */
class BscScan {
  apikey: string | undefined;
  baseUrl: string;

  accounts: Accounts;
  contracts: Contracts;
  transactions: Transactions;
  stats: Stats;

  /**
   * Creates an instance of BscScan.
   * @param {BscScanOptions} opts
   * @memberof BscScan
   */
  constructor(opts: BscScanOptions) {
    const { apikey } = opts;

    this.apikey = apikey;
    this.baseUrl = opts.baseUrl || BASE_URL;

    this.accounts = new Accounts(this);
    this.contracts = new Contracts(this);
    this.stats = new Stats(this);
    this.transactions = new Transactions(this);
  }

  /**
   * Creates a new HTTP request
   *
   * @param {string} [method="GET"]
   * @param {string} module
   * @param {string} action
   * @param {Record<string, any>} [opts={}]
   * @returns {Request}
   * @memberof BscScan
   */
  newRequest(method = "GET", module: string, action: string, opts: Record<string, any> = {}): Request {
    const params = new URLSearchParams({ module, action });

    if (this.apikey) {
      params.set("apikey", this.apikey);
    }

    for (const [key, value] of Object.entries(opts)) {
      if (Array.isArray(value)) {
        params.set(key, value.join(","));
      } else {
        params.set(key, value);
      }
    }

    const headers: Headers = new Headers();
    headers.set("Content-Type", "application/json; charset=UTF-8");

    const init: RequestInit = { method, headers };

    const url = new URL("/api?" + params.toString(), baseUrl);

    const request = new Request(url.toString(), init);
    return request;
  }

  /**
   * Executes a simplified API query
   *
   * @param {string} module
   * @param {string} action
   * @param {Record<string, any>} [opts={}]
   * @returns {Promise<any>}
   * @memberof BscScan
   */
  async query(module: string, action: string, opts: Record<string, any> = {}): Promise<any> {
    return this.do(this.newRequest("GET", module, action, opts));
  }

  /**
   * Executes an HTTP request
   *
   * @template T
   * @param {Request} request
   * @returns {Promise<T>}
   * @memberof BscScan
   */
  async do<T>(request: Request): Promise<T> {
    const response: Response = await fetch(request);
    const responseBody: APIResponse<T> = await response.json();

    if (responseBody.status === "0") {
      throw new Error(responseBody.message + " - " + responseBody.result);
    }

    return responseBody.result;
  }
}

export { BscScan };
