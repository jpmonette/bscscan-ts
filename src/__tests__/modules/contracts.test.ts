import "isomorphic-fetch";
import { BscScan } from "../../client";
import { VerifySourceCodeRequest } from "../../typings";
import { server } from "../server";
import { API_KEY } from "../setup";

describe("contracts", () => {
  const client = new BscScan({ apikey: API_KEY });

  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  it("should getAbi", async () => {
    const opts = { address: "0x0e09fabb73bd3ade0a17ecc321fd13a19e81ce82" };
    expect(await client.contracts.getAbi(opts)).toMatchSnapshot();
  });

  it("should getSourceCode", async () => {
    const opts = { address: "0x0e09fabb73bd3ade0a17ecc321fd13a19e81ce82" };
    expect(await client.contracts.getSourceCode(opts)).toMatchSnapshot();
  });

  it("should verifySourceCode", async () => {
    const opts: VerifySourceCodeRequest = {
      contractAddress: "0x0e09fabb73bd3ade0a17ecc321fd13a19e81ce82",
      contractName: "CAKE",
      sourceCode: "",
      compilerVersion: "v0.6.12+commit.27d51765",
      optimizationUsed: "1",
      runs: 5000,
      licenseType: "None",
    };
    expect(await client.contracts.verifySourceCode(opts)).toMatchSnapshot();
  });

  it("should verifyProxyContract", async () => {
    const opts = {
      address: "0xbc46363a7669f6e12353fa95bb067aead3675c29",
      expectedImplementation: "0xe45a5176bc0f2c1198e2451c4e4501d4ed9b65a6",
    };
    expect(await client.contracts.verifyProxyContract(opts)).toMatchSnapshot();
  });

  it("should checkProxyVerification", async () => {
    const opts = { guid: "gwgrrnfy56zf6vc1fljuejwg6pelnc5yns6fg6y2i6zfpgzquz" };
    expect(await client.contracts.checkProxyVerification(opts)).toMatchSnapshot();
  });
});
