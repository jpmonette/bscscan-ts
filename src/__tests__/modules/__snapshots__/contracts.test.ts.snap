// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contracts should checkProxyVerification 1`] = `"0xcbdcd3815b5f975e1a2c944a9b2cd1c985a1cb7f"`;

exports[`contracts should getAbi 1`] = `"[]"`;

exports[`contracts should getSourceCode 1`] = `
Array [
  Object {
    "ABI": "[]",
    "CompilerVersion": "v0.6.12+commit.27d51765",
    "ConstructorArguments": "",
    "ContractName": "CakeToken",
    "EVMVersion": "Default",
    "Implementation": "",
    "Library": "",
    "LicenseType": "None",
    "OptimizationUsed": "1",
    "Proxy": "0",
    "Runs": "5000",
    "SourceCode": "pragma solidity 0.6.12;",
    "SwarmSource": "ipfs://9755240809e31aec9fa5790314682233ca431b7c4f252d7e4bba347e2e742086",
  },
]
`;

exports[`contracts should verifyProxyContract 1`] = `"a7xygt433w43um5fda3m6d9jk62jwtjhqdr9zcwj9fmfzcj14r"`;

exports[`contracts should verifySourceCode 1`] = `"a7xygt433w43um5fda3m6d9jk62jwtjhqdr9zcwj9fmfzcj14r"`;
